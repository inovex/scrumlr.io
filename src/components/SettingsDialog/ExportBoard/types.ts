import {Color} from "constants/colors";

export interface BoardType {
  id: string;
  name: string;
  showAuthors: boolean;
  showVoting: string;
  accessPolicy: string;
}

export interface ColumnType {
  id: string;
  name: string;
  color: Color;
  visible: boolean;
}

export interface NoteType {
  id: string;
  author: string;
  text: string;
  position: {
    column: string;
    stack: string | null;
    rank: number;
  };
}

export interface ParticipantType {
  role: string;
  user: {
    id: string;
    name: string;
  };
}

export interface VotingType {
  votes: {
    votes: number;
    votesPerNote: [];
  };
}

export interface BoardDataType {
  board: BoardType;
  columns: ColumnType[];
  notes: NoteType[];
  participants: ParticipantType[];
  votings: VotingType[];
}
