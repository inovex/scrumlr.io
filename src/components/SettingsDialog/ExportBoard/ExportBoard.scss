@import "src/constants/style.scss";

.export-board__button-reverse-order {
  flex-direction: row-reverse;
  justify-content: flex-end;
}

.export-board__button-print-view {
  svg {
    height: $icon--medium;
  }
}

[theme="dark"] {
  .export-board__button-reverse-order svg g g {
    opacity: 1;
  }
  .export-board__button-reverse-order svg text {
    fill: $color-white;
    stroke: $color-white;
  }
}
