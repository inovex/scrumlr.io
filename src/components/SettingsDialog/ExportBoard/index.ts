export * from "./ExportBoard";
export * from "./ExportHintHiddenColumns";
