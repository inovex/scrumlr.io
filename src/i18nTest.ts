import {join} from "path";
import i18n from "i18next";
import {initReactI18next} from "react-i18next";
import Backend from "i18next-fs-backend";

i18n
  .use(Backend)
  .use(initReactI18next)
  .init({
    backend: {
      loadPath: join(__dirname, "../public/locales/en/translation.json"),
    },
    initImmediate: false,
    lng: "en",
    debug: false,
    react: {
      useSuspense: false,
    },
  });

export default i18n;
