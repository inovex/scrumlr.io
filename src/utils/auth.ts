import {Toast} from "utils/Toast";
import {API} from "api";
import i18n from "i18next";
import store from "store";
import {Actions} from "store/action";
import {SERVER_HTTP_URL} from "../config";

/**
 * Sign in anonymously.
 *
 * @param displayName Display name of the parse auth user.
 *
 * @returns Promise with user credentials on successful sign in, null otherwise.
 */
const signInAnonymously = async (displayName: string) => {
  try {
    const user = await API.signInAnonymously(displayName);
    if (user) {
      store.dispatch(Actions.signIn(user.id, user.name));
    }
    return true;
  } catch (err) {
    Toast.error({title: i18n.t("Toast.authenticationError")});
    return null;
  }
};

/**
 * Redirects to OAuth page of provider:
 *    https://accounts.google.com/o/oauth2/v2/auth/oauthchooseaccount...
 *    https://github.com/login/oauth/authorize...
 *    https://login.microsoftonline.com/common/oauth2/v2.0/authorize...
 * @param authProvider name of the OAuth Provider
 * @param originURL origin URL
 */
const signInWithAuthProvider = async (authProvider: string, originURL: string) => {
  window.location.href = `${SERVER_HTTP_URL}/login/${authProvider}?state=${encodeURIComponent(originURL)}`;
};

export const Auth = {
  signInAnonymously,
  signInWithAuthProvider,
};
