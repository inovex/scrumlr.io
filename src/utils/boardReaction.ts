import {BoardReactionType} from "types/boardReaction";

// these are custom events that can be emitted and subscribed to.
// using redux is overkill, so we decided to separate the logic, similar to toasts
export const ADD_BOARD_REACTION = "scrumlr.io/addBoardReaction";
export const REMOVE_BOARD_REACTION = "scrumlr.io/removeBoardReaction";

const add = (reaction: BoardReactionType) => {
  const addBoardReactionEvent = new CustomEvent<BoardReactionType>(ADD_BOARD_REACTION, {detail: reaction});
  document.dispatchEvent(addBoardReactionEvent);
};

const remove = (id: string) => {
  const removeBoardReactionEvent = new CustomEvent<string>(REMOVE_BOARD_REACTION, {detail: id});
  document.dispatchEvent(removeBoardReactionEvent);
};

export const BoardReactionEmitter = {
  add,
  remove,
};
