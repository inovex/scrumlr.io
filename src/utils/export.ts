import {saveAs} from "file-saver";
import {BoardDataType, BoardType, ColumnType, NoteType, ParticipantType, VotingType} from "components/SettingsDialog/ExportBoard/types";
import i18n from "i18next";
import {DEFAULT_BOARD_NAME, DEFAULT_URL} from "constants/misc";
import {API} from "../api";

const {t} = i18n;

export const fileName = (name?: string) => {
  const date = new Date().toJSON().slice(0, 10);
  return `${date}_${name}`;
};

export const exportAsCSV = async (id: string, name?: string) => {
  const response = await API.exportBoard(id, "text/csv");
  const blob = await response.blob();
  saveAs(blob, `${fileName(name ?? DEFAULT_BOARD_NAME)}.csv`);
};

export const exportAsJSON = async (id: string, name?: string) => {
  const response = await API.exportBoard(id, "application/json");
  const json = await response.json();
  const blob = new Blob([JSON.stringify(json)], {type: "application/json"});
  saveAs(blob, `${fileName(name ?? DEFAULT_BOARD_NAME)}.json`);
};

export const getNoteVotes = (noteId: string, votings: VotingType[]) => votings[0].votes?.votesPerNote[noteId]?.total ?? 0;

export const compareNotes = (a: {id: string; position: {rank: number}}, b: {id: string; position: {rank: number}}, votings: VotingType[]) => {
  if (votings && getNoteVotes(a.id, votings) !== getNoteVotes(b.id, votings)) {
    return getNoteVotes(a.id, votings) > getNoteVotes(b.id, votings) ? -1 : 1;
  }
  return a.position.rank > b.position.rank ? -1 : 1;
};

export const getChildNotes = (notes: NoteType[], votings: VotingType[], noteId: string) =>
  notes.filter((n) => n.position.stack === noteId).sort((a, b) => compareNotes(a, b, votings));

export const getAuthorName = (authorId: string, participants: ParticipantType[]) => participants.filter((p: ParticipantType) => p.user?.id === authorId)[0].user?.name;

const mdItalicBrackets = (addBrackets: boolean, input: string) => (addBrackets ? `_(${input})_` : input);

const mdBoardHeader = (boardName: string) => `# ${boardName}\n\n`;

const mdBoardProperties = (board: BoardType, participants: ParticipantType[]) => {
  const linkToBoard = `- [${t("MarkdownExport.linkToBoard")}](${window.location.href.replace("/settings/export", "")})\n`;
  const participantsNumber = `- ${t("MarkdownExport.participants")}: ${participants.length}\n`;
  const authorsHidden = !board.showAuthors ? `- ${t("MarkdownExport.showAuthorsDisabled")}\n` : "";
  return `${participantsNumber + authorsHidden + linkToBoard}\n`;
};

const mdVotesPerNote = (noteId: string, votings: VotingType[]) => {
  const votes = votings ? getNoteVotes(noteId, votings) : 0;
  return votes ? `${votes} ${votes === 1 ? t("MarkdownExport.vote") : t("MarkdownExport.votes")}` : "";
};

const mdNote = (note: NoteType, author: string, votes: string, stack: string) => {
  const nested = stack === "" && note.position.stack;
  const bracketsRequired = !!author || !!votes;
  const separator = !!author && !!votes ? ", " : "";
  return `${nested ? "\n    " : "\n"}- ${note.text} ${mdItalicBrackets(bracketsRequired, `${author + separator + votes}`)}${stack}`;
};

const mdStack = (notesInStack: NoteType[], boardData: BoardDataType) =>
  notesInStack.length > 0
    ? `${notesInStack
        .map((n) =>
          mdNote(
            n,
            boardData.board.showAuthors ? getAuthorName(n.author, boardData.participants) : "",
            boardData.board.showAuthors ? mdVotesPerNote(n.id, boardData.votings) : "",
            ""
          )
        )
        .join("")}`
    : "";

const mdNotesPerColumn = (columnId: string, boardData: BoardDataType) =>
  boardData.notes
    .filter((n) => !n.position.stack && n.position.column === columnId)
    .sort((a, b) => compareNotes(a, b, boardData.votings))
    .map((n) => {
      const votes = boardData.board.showVoting ? mdVotesPerNote(n.id, boardData.votings) : "";
      const author = boardData.board.showAuthors ? getAuthorName(n.author, boardData.participants) : "";
      const stack = mdStack(getChildNotes(boardData.notes, boardData.votings, n.id), boardData);
      return mdNote(n, author, votes, stack);
    })
    .join("");

const mdColumns = (boardData: BoardDataType) => {
  const columnList = boardData.columns
    .filter((c) => boardData.notes.filter((n) => n.position.column === c.id).length > 0)
    .map((c: ColumnType) => `## ${c.name} (${boardData.notes.filter((n) => n.position.column === c.id).length} ${t("MarkdownExport.notes")})\n${mdNotesPerColumn(c.id, boardData)}`)
    .join("\n\n");
  return `${columnList}\n\n`;
};

const mdBranding = () =>
  `> ${t("PrintView.GeneratedOn")} [scrumlr.io](${DEFAULT_URL})  \n${t(
    "PrintView.ProvidedBy"
  )} [inovex](https://www.inovex.de)  \n\n![Scrumlr Logo](${DEFAULT_URL}/scrumlr-logo-light.svg)`;

const mdTemplate = (boardData: BoardDataType) =>
  mdBoardHeader(boardData.board.name || DEFAULT_BOARD_NAME) + mdBoardProperties(boardData.board, boardData.participants) + mdColumns(boardData) + mdBranding();

export const getMarkdownExport = async (id: string) => {
  const response = await API.exportBoard(id, "application/json");
  const json: BoardDataType = await response.json();
  return `${mdTemplate(json)}`;
};
