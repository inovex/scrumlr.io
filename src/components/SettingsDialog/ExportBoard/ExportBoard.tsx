import classNames from "classnames";
import {VFC} from "react";
import {useTranslation} from "react-i18next";
import {ReactComponent as ExportCSV} from "assets/icon-export-csv.svg";
import {ReactComponent as ExportJSON} from "assets/icon-export-json.svg";
import {ReactComponent as PrintIcon} from "assets/icon-print.svg";
import {ReactComponent as ClipboardIcon} from "assets/icon-clipboard.svg";
import {useAppSelector} from "store";
import {exportAsJSON, exportAsCSV, getMarkdownExport} from "utils/export";
import {Toast} from "utils/Toast";
import {TOAST_TIMER_SHORT} from "constants/misc";
import {SettingsButton} from "../Components/SettingsButton";
import {ExportHintHiddenColumns} from "./ExportHintHiddenColumns";
import "./ExportBoard.scss";
import "../SettingsDialog.scss";

export const ExportBoard: VFC = () => {
  const {t} = useTranslation();

  const boardId = useAppSelector((state) => state.board.data!.id);
  const boardName = useAppSelector((state) => state.board.data!.name);
  const columns = useAppSelector((state) => state.columns);

  return (
    <div data-testid="export" className="settings-dialog__container">
      <div className="settings-dialog__header">
        <h2 className={classNames("settings-dialog__header-text", "accent-color__backlog-blue")}> {t("ExportBoardOption.title")}</h2>
      </div>

      <div className={classNames("settings-dialog__group", "accent-color__backlog-blue")}>
        <SettingsButton
          label={t("ExportBoardOption.exportAsJson")}
          icon={ExportJSON}
          className="export-board__button-reverse-order"
          onClick={() => {
            exportAsJSON(boardId, boardName);
          }}
          data-testid="export-json"
        />
        <hr className="settings-dialog__separator" />
        <SettingsButton
          label={t("ExportBoardOption.exportAsCSV")}
          icon={ExportCSV}
          className="export-board__button-reverse-order"
          onClick={() => {
            exportAsCSV(boardId, boardName);
          }}
          data-testid="export-csv"
        />
        <hr className="settings-dialog__separator" />
        <SettingsButton
          label={t("ExportBoardOption.openPrintView")}
          icon={PrintIcon}
          className="export-board__button-reverse-order export-board__button-print-view"
          onClick={() => {
            window.open(`/board/${boardId}/print`, "_blank");
          }}
        />
        <hr className="settings-dialog__separator" />
        <SettingsButton
          label={t("ExportBoardOption.exportToClipboard")}
          icon={ClipboardIcon}
          className="export-board__button-reverse-order"
          onClick={() => {
            getMarkdownExport(boardId).then((result) => {
              navigator.clipboard.writeText(result).then(() => {
                Toast.success({title: t("ExportBoardOption.copyToClipboardSuccess"), autoClose: TOAST_TIMER_SHORT});
              });
            });
          }}
          data-testid="export-markdown"
        />
      </div>

      <ExportHintHiddenColumns columns={columns} />
    </div>
  );
};
