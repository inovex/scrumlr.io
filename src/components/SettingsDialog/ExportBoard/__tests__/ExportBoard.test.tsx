import {fireEvent, render, waitFor} from "@testing-library/react";
import {Provider} from "react-redux";
import {exportAsJSON, exportAsCSV} from "utils/export";
import {ExportBoard} from "../ExportBoard";
import {I18nextProvider} from "react-i18next";
import i18nTest from "i18nTest";
import getTestStore from "utils/test/getTestStore";
import {BrowserRouter} from "react-router-dom";

jest.mock("utils/export", () => ({
  ...jest.requireActual("utils/export"),
  exportAsJSON: jest.fn(),
  exportAsCSV: jest.fn(),
}));

const renderExportBoard = () => {
  return render(
    <I18nextProvider i18n={i18nTest}>
      <Provider store={getTestStore()}>
        <BrowserRouter>
          <ExportBoard />
        </BrowserRouter>
      </Provider>
    </I18nextProvider>
  );
};

describe("Export Boad Tests", () => {
  test("render ExportBoard correctly", async () => {
    const exportBoard = renderExportBoard();
    expect(exportBoard).toMatchSnapshot();
  });

  test("json function called", async () => {
    const exportBoard = renderExportBoard();
    expect(exportBoard.getByTestId("export-json")).not.toBeNull();

    const button = exportBoard.getByTestId("export-json")!;
    fireEvent.click(button);

    await waitFor(() => {
      expect(exportAsJSON).toBeCalled();
    });
  });

  test("csv function called", async () => {
    const exportBoard = renderExportBoard();
    expect(exportBoard.getByTestId("export-csv")).not.toBeNull();

    const button = exportBoard.getByTestId("export-csv")!;
    fireEvent.click(button);

    await waitFor(() => {
      expect(exportAsCSV).toBeCalled();
    });
  });
});
