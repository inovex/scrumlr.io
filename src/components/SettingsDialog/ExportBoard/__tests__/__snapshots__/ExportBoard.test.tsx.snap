// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Export Boad Tests render ExportBoard correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="settings-dialog__container"
        data-testid="export"
      >
        <div
          class="settings-dialog__header"
        >
          <h2
            class="settings-dialog__header-text accent-color__backlog-blue"
          >
             
            Export board
          </h2>
        </div>
        <div
          class="settings-dialog__group accent-color__backlog-blue"
        >
          <button
            aria-label="Export as JSON"
            class="settings-option-button export-board__button-reverse-order"
            data-testid="export-json"
            type="button"
          >
            <span
              class="settings-option-button__label"
            >
              Export as JSON
            </span>
            <svg
              class="settings-option-button__icon"
            >
              icon-export-json.svg
            </svg>
          </button>
          <hr
            class="settings-dialog__separator"
          />
          <button
            aria-label="Export as CSV"
            class="settings-option-button export-board__button-reverse-order"
            data-testid="export-csv"
            type="button"
          >
            <span
              class="settings-option-button__label"
            >
              Export as CSV
            </span>
            <svg
              class="settings-option-button__icon"
            >
              icon-export-csv.svg
            </svg>
          </button>
          <hr
            class="settings-dialog__separator"
          />
          <button
            aria-label="Open print view"
            class="settings-option-button export-board__button-reverse-order export-board__button-print-view"
            type="button"
          >
            <span
              class="settings-option-button__label"
            >
              Open print view
            </span>
            <svg
              class="settings-option-button__icon"
            >
              icon-print.svg
            </svg>
          </button>
          <hr
            class="settings-dialog__separator"
          />
          <button
            aria-label="Copy Markdown to Clipboard"
            class="settings-option-button export-board__button-reverse-order"
            data-testid="export-markdown"
            type="button"
          >
            <span
              class="settings-option-button__label"
            >
              Copy Markdown to Clipboard
            </span>
            <svg
              class="settings-option-button__icon"
            >
              icon-clipboard.svg
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="settings-dialog__container"
      data-testid="export"
    >
      <div
        class="settings-dialog__header"
      >
        <h2
          class="settings-dialog__header-text accent-color__backlog-blue"
        >
           
          Export board
        </h2>
      </div>
      <div
        class="settings-dialog__group accent-color__backlog-blue"
      >
        <button
          aria-label="Export as JSON"
          class="settings-option-button export-board__button-reverse-order"
          data-testid="export-json"
          type="button"
        >
          <span
            class="settings-option-button__label"
          >
            Export as JSON
          </span>
          <svg
            class="settings-option-button__icon"
          >
            icon-export-json.svg
          </svg>
        </button>
        <hr
          class="settings-dialog__separator"
        />
        <button
          aria-label="Export as CSV"
          class="settings-option-button export-board__button-reverse-order"
          data-testid="export-csv"
          type="button"
        >
          <span
            class="settings-option-button__label"
          >
            Export as CSV
          </span>
          <svg
            class="settings-option-button__icon"
          >
            icon-export-csv.svg
          </svg>
        </button>
        <hr
          class="settings-dialog__separator"
        />
        <button
          aria-label="Open print view"
          class="settings-option-button export-board__button-reverse-order export-board__button-print-view"
          type="button"
        >
          <span
            class="settings-option-button__label"
          >
            Open print view
          </span>
          <svg
            class="settings-option-button__icon"
          >
            icon-print.svg
          </svg>
        </button>
        <hr
          class="settings-dialog__separator"
        />
        <button
          aria-label="Copy Markdown to Clipboard"
          class="settings-option-button export-board__button-reverse-order"
          data-testid="export-markdown"
          type="button"
        >
          <span
            class="settings-option-button__label"
          >
            Copy Markdown to Clipboard
          </span>
          <svg
            class="settings-option-button__icon"
          >
            icon-clipboard.svg
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
